(DOCSP-26118): updated bundle a realm page (#2528)
mongodb/docs-realm: source/sdk/flutter/realm-database/realm-files/bundle.txt

   You can also add data to your realm the first time an application opens it
   using the :ref:`initial data callback function <flutter-initial-data-callback>`.

.. important:: Bundling Synced Realms

   If your backend application has enabled :ref:`advanced backend compaction

>    <advanced-backend-compaction>` by configuring a

Reviewer: "Compaction" is a PBS-only concept. In Flexible Sync it is "Trimming" which is much more efficient

   :ref:`client maximum offline time <client-maximum-offline-time>`,
   users could experience a client reset the first time they open the
   bundled realm file. This can happen if the bundled realm file was
   generated more than **client maximum offline time** days before the user syncs

>    the realm for the first time.

Reviewer: I would make sure people know that the benefits of trimming far outweigh the benefits of using the bundled realm in most cases. The bundled realm approach really works best in tandem with trimming for customers who have a complete ownership of the lifecycle of their app. IE, they can release their app, their users are their employees, and they all update it automatically. This is because they will need to re-release their app with a new bundled realm file every MaxOfflineTime days (which would not be ideal I imagine)

   Applications experiencing a client reset will download the full state of the
   realm from the application backend. This negates the
   advantages of bundling a realm file. To prevent client resets and
   preserve the advantages of realm file bundling:

>    - Avoid using a client maximum offline time in applications that

Reviewer 2: Perhaps ask someone from the backend, like @tkaye407 if we really want to advocate disabling maximum offline time. It will allow history to grow unhindered.

In general I think we should really warn against using bundled realm files with flexible sync.. the footgun is big, and the benefit much less than previously given recent optimizations of bootstrap.

         :language: dart
         :caption: lib/main.dart

.. _flutter-bundle-synced-realm:

Bundle a Synced Realm
---------------------

In most cases, you should **not** bundle a synced realm. If the bundled realm
was last updated further in the past than :ref:`client maximum offline time
<client-maximum-offline-time>`, the user experiences a client reset
the first time they open the bundled realm file. The client reset causes
the application to download the full state of the realm from the application backend.
This negates the advantages of bundling a realm file.

Rather than bundling a synced realm, you can bootstrap application data using
sync subscriptions. To learn more, refer to :ref:`Manage Sync Subscriptions
<flutter-flexible-sync-manage-subscriptions>`.

You should only bundle a synced realm if your use case meets the following criteria:

- You can ensure that the users have a version of the app with the bundled synced realm
  that is less old than the client maximum offline time.
- The initial bundled data is very large **and** app is being used in a situation
  with limited internet bandwidth, so an initial data download using sync subscriptions

>   would take too long.

Author: question for technical reviewer(s): does this capture the use case well?

      .. literalinclude:: /examples/generated/flutter/main.snippet.include-bundled-realm.dart
         :language: dart
         :caption: lib/main.dart

> 

Reviewer 3: I can't comment above since it wasn't edited in this PR, but given the convos here and the recommendation not to bundle a synced realm in most cases, you probably want to update the intro paragraph for the page to add something around that. Currently, in the intro paragraph, it's not obvious that this statement: 

> You might want to seed your mobile app with some initial data that will be available to users on the initial launch of the app.

Pertains only to non-synced realms. I see the section about synced realms in the "On This Page" ToC, but we've had some feedback from docs visitors that they ignore that section, so I think we should be explicit in the intro - something like:

> If your app uses a synced realm, you may not want to bundle it. For more information, refer to...

.. _flutter-bundle-synced-realm:

Bundle a Synced Realm
---------------------

In most cases, you should **not** bundle a synced realm. If the bundled realm
was last updated further in the past than :ref:`client maximum offline time
<client-maximum-offline-time>`, the user experiences a client reset
the first time they open the bundled realm file. The client reset causes
the application to download the full state of the realm from the application backend.
This negates the advantages of bundling a realm file.

> Rather than bundling a synced realm, you can bootstrap application data using

Reviewer 3: What does "bootstrap application data using sync subscriptions" mean? In the linked page, we do not mention the word boostrap at all. The way we refer to it here, though, it makes me think there's something special to do relative to this.

Author: makes sense. will rephrase

sync subscriptions. To learn more, refer to :ref:`Manage Sync Subscriptions
<flutter-flexible-sync-manage-subscriptions>`.

You should only bundle a synced realm if your use case meets the following criteria:

- You can ensure that the users have a version of the app with the bundled synced realm

>   that is less old than the client maximum offline time.

Reviewer 3: Nit: "less old than" doesn't seem quite right here. I'm not sure my suggestion is much better, but I think we should somehow improve this wording?
```suggestion
  that was created more recently than the client maximum offline time.
```

> - The initial bundled data is very large **and** app is being used in a situation

Reviewer 3: Suggest adding an article here, but also - does this have to be _and_? Isn't it a valid use case to want to bundle a realm when an app is being used in a situation with limited bandwidth regardless of the size of the bundled data - i.e. _or_ when bandwidth is limited?

```suggestion
- The initial bundled data is very large **and** the app is being used in a situation
```

  with limited internet bandwidth, so an initial data download using sync subscriptions
  would take too long.
- All app users have backend permission to view the data included in the bundle.
  If a user doesn't have permission to view this data, it will be removed from
  their device when the realm sync with Atlas via a :ref:`compensating write error

>   <flutter-compensating-writes>`.

Author: question for technical reviewer(s): does this capture the use case well?

To bundle a synced realm, perform the following:

> #. Connect to your App and authenticate a user to use Device Sync.

Reviewer 3: The mix of "App" and "Device Sync" here make me think that this "App" instance refers to Atlas App Services App. As this is the first usage on the page, I think the marketing guidance is to use the full product name.

If we're referring to the Flutter `App` class here, maybe we should use monospace or link out to the API to be clear that's what we mean? And if so, I might recommend linking to the "Authenticate Users" page and omit the "to use Device Sync" part.

> #. Add a subscription to the realm. You need a subscription to write to a synced realm.

Reviewer 3: Should this bullet link to the "Manage Sync Subscriptions" page?

Author: not linking the first 4 bullet points to keep instructions more focused on the bundling part.

#. Add Data to the synced realm.

> #. Sync all changes with the Device Sync server.

Reviewer 3: "Sync all changes with the Device Sync server" as a step here makes me think there's something active we need to do to sync changes. If we mean wait for changes to sync, maybe we want to link to or refer to the specific API methods for that? 

(Also, it occurs to me that the "Write Data to a Synced Realm" page should probably have a section about the "waitForUpload"/"waitForDownload" APIs. I know we intended that page to be about compensating writes, but as a reader, if I see "Write to a Synced Realm" - I would expect to see those APIs called out there specifically. We show them in an example but don't highlight them in the "On This Page.")

Author: > (Also, it occurs to me that the "Write Data to a Synced Realm" page should probably have a section about the "waitForUpload"/"waitForDownload" APIs.

agree. already planning on refreshing that page to look more like the swift one. will include this in the update too. 

created ticket for this change - [DOCSP-27850](https://jira.mongodb.org/browse/DOCSP-27850)

Author: > "Sync all changes with the Device Sync server" as a step here makes me think there's something active we need to do to sync changes. If we mean wait for changes to sync, maybe we want to link to or refer to the specific API methods for that?

rephrasing this line
